ctru-rs/src/services/uds.rs

```diff
@@ -0,0 +1,1273 @@
+pub struct NetworkScanInfo {
+    /// NWM output structure.
+    pub datareply_entry: ctru_sys::nwmBeaconDataReplyEntry,
+    /// Information about the network.
+    pub network: ctru_sys::udsNetworkStruct,
+    /// All nodes on the network (first node is the server,
+    /// max 16, `None` means no node connected).
+    pub nodes: [Option<NodeInfo>; 16],
+}
+
+impl From<ctru_sys::udsNetworkScanInfo> for NetworkScanInfo {
+    fn from(value: ctru_sys::udsNetworkScanInfo) -> Self {
+        Self {
+            datareply_entry: value.datareply_entry,
+            network: value.network,
+            nodes: value.nodes.map(|n| {
+                if n.uds_friendcodeseed != 0 {
+                    Some(n.into())
+                } else {
+                    None
+                }
+            }),
+        }
+    }
+}
+
+/// Status of the connection.
+#[doc(alias = "udsConnectionStatus")]
+#[derive(Debug)]
+pub struct ConnectionStatus {
+    /// Raw status information
+    // TODO: is this in some kind of readable format?
+    pub status: u32,
+    unk_x4: u32,
+    /// Network node ID for the current device.
+    pub cur_node_id: NodeID,
+    unk_xa: u16,
+    unk_xc: [u32; 8],
+    /// Number of nodes connected to the network.
+    pub total_nodes: u8,
+    /// Maximum nodes allowed on this network.
+    pub max_nodes: u8,
+    /// Bitmask for which of the 16 possible nodes are connected
+    /// to this network; bit 0 is the server, bit 1 is the first
+    /// original client, etc.
+    pub node_bitmask: u16,
+}
+
+impl From<ctru_sys::udsConnectionStatus> for ConnectionStatus {
+    fn from(value: ctru_sys::udsConnectionStatus) -> Self {
+        Self {
+            status: value.status,
+            unk_x4: value.unk_x4,
+            cur_node_id: value
+                .cur_NetworkNodeID
+                .try_into()
+                .expect("UDS service should always provide a valid NetworkNodeID"),
+            unk_xa: value.unk_xa,
+            unk_xc: value.unk_xc,
+            total_nodes: value.total_nodes,
+            max_nodes: value.max_nodes,
+            node_bitmask: value.node_bitmask,
+        }
+    }
+}
+
+/// Status of the service handle.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ServiceStatus {
+    /// Not connected to or hosting a network.
+    Disconnected,
+    /// Connected to a network.
+    Client,
+    /// Hosting a network.
+    Server,
+}
+
+/// Handle to the UDS service.
+pub struct Uds {
+    _service_handler: ServiceReference,
+    context: Option<ctru_sys::udsBindContext>,
+    network: Option<ctru_sys::udsNetworkStruct>,
+    scan_buf: Box<[u8; Self::SCAN_BUF_SIZE]>,
+}
+
+static UDS_ACTIVE: Mutex<()> = Mutex::new(());
+
+impl Uds {
+    /// Size of one frame.
+    const RECV_FRAME_SIZE: usize = ctru_sys::UDS_DATAFRAME_MAXSIZE as usize;
+
+    /// Size of receive buffer; max frame size * 8.
+    const RECV_BUF_SIZE: u32 = ctru_sys::UDS_DEFAULT_RECVBUFSIZE;
```

Comment:
There's no good reason (as far as I'm aware) to have the user able to change them, but they're properties of the data structure, rather than something you'd want to limit within a single function.